pkg/e2e/autoscaler/autoscaler.go

```diff
@@ -176,166 +139,246 @@ func machineAutoscalerResource(targetMachineSet *mapiv1beta1.MachineSet, minRepl
+			defer maxNodesTotalStartTimeLock.Unlock()
+			t := metav1.Now()
+			maxNodesTotalStartTime = &t
+		})
+		g.By(fmt.Sprintf("Waiting for MaxNodesTotalReached event. Current count #%v", atomic.LoadUint64(&maxNodesTotalReachedCount)))
+		return atomic.LoadUint64(&maxNodesTotalReachedCount) >= 1, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
 
-		g.By("Creating workload")
-		err = client.Create(context.TODO(), newWorkLoad())
-		o.Expect(err).NotTo(o.HaveOccurred())
+	//
+	// Scale down.
+	//
 
-		scaleUpFunc(client, targetMachineSet1, nodeTestLabel1, initialNumberOfReplicas1, initialNumberOfReplicas1+2)
-		scaleUpFunc(client, targetMachineSet0, nodeTestLabel0, initialNumberOfReplicas0, initialNumberOfReplicas1+1)
+	var scaleDownStartTimeLock sync.Mutex
+	var scaleDownStartTime *metav1.Time
+	var scaleDownCount = uint64(clusterExpansionSize - 1)
 
-		// Delete workload
-		g.By("Deleting workload")
-		err = e2e.DeleteObjectsByLabels(context.TODO(), client, map[string]string{autoscalingTestLabel: ""}, &batchv1.JobList{})
-		o.Expect(err).NotTo(o.HaveOccurred())
+	matchScaleDownEvent := func(event *corev1.Event) {
+		scaleDownStartTimeLock.Lock()
+		defer scaleDownStartTimeLock.Unlock()
+		if scaleDownStartTime == nil || event.CreationTimestamp.Before(scaleDownStartTime) {
+			return
+		}
+		if isAutoscalerScaleDownEvent(event) {
+			atomic.AddUint64(&scaleDownCount, ^uint64(0))
+		}
+	}
 
-		// As we have just deleted the workload the autoscaler will
-		// start to scale down the unneeded nodes. We wait for that
-		// condition; if successful we assert that (a smoke test of)
-		// scale down is functional.
-		scaleDownFunc(client, targetMachineSet0.Name, nodeTestLabel0, initialNumberOfReplicas0)
-		scaleDownFunc(client, targetMachineSet1.Name, nodeTestLabel1, initialNumberOfReplicas1)
+	// By deleting the workload the autoscaler will start to scale
+	// down the nodes it just scaled up. We wait for that
+	// condition; if successful we assert that (a smoke test of)
+	// scale down is functional.
+
+	var scaleDownOnce sync.Once
+	err = wait.PollImmediate(3*time.Second, e2e.WaitLong, func() (bool, error) {
+		scaleDownOnce.Do(func() {
+			scaleDownStartTimeLock.Lock()
+			defer scaleDownStartTimeLock.Unlock()
+			t := metav1.Now()
+			scaleDownStartTime = &t
+
+			matchEventFuncLock.Lock()
+			defer matchEventFuncLock.Unlock()
+			matchEventFunc = matchScaleDownEvent
+
+			g.By("Deleting workload")
+			err := e2e.DeleteObjectsByLabels(context.TODO(), client, map[string]string{autoscalingTestLabel: ""}, &batchv1.JobList{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			if len(cleanupObjects) > 1 && cleanupObjects[len(cleanupObjects)-1] == workload {
+				// workload already deleted above
+				cleanupObjects = cleanupObjects[:len(cleanupObjects)-1]
+			}
+		})
+		g.By(fmt.Sprintf("Waiting for cluster to scale down %d more nodes", atomic.LoadUint64(&scaleDownCount)))
+		return atomic.LoadUint64(&scaleDownCount) == 0, nil
```

Comment:
Why scaleDownCount should be 0? clusterExpansionSize has master machines + worker nodes + (extra scaled out - 1). Why it is being asserted that all initial worker nodes and master nodes will also be scaled down?